### PR TITLE
llama : support InternLM2

### PR DESCRIPTION
**Pull request by SolenoidWGT**
Dear llama.cpp developer @ggerganov:

Hello, this PR is to support the [InternLM2](https://huggingface.co/internlm/internlm2-20b). InternLM2 is a GQA-based transform architecture and offers effective support for long contexts of up to 200,000 characters. Now more and more users want to explore the capabilities of InternLM2 using their familiar llama.cpp API.  We sincerely hope that our PR can be accepted. Thank you.

For information about InternLM2, please visit https://github.com/InternLM/InternLM, 

related issue: https://github.com/ggerganov/llama.cpp/issues/3133, https://github.com/ggerganov/llama.cpp/issues/3551, https://github.com/ggerganov/llama.cpp/issues/4360, https://github.com/ggerganov/llama.cpp/issues/5031

cc: @yhcc, @vansinhu, @sunpengsdu

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -11220,7 +11409,8 @@ int32_t llama_tokenize(
                      int32_t   n_max_tokens,
                         bool   add_bos,
                         bool   special) {
-    auto res = llama_tokenize_internal(model->vocab, std::string(text, text_len), add_bos, special);
+    bool add_space_prefix = false ? model->arch == LLM_ARCH_INTERNLM2 : true;
```
Comment (ggerganov) [inline llama.cpp:11412]: The condition is wrong:

```suggestion
    bool add_space_prefix = model->arch == LLM_ARCH_INTERNLM2 ? true : false;
```

Is this change necessary? Seems strange to add prefix just for a specific model architecture. Can this be fixed outside of `llama.cpp`?

**Comment by cmp-nct** [comment]
> Dear llama.cpp developer @ggerganov:
> 
> Hello, this PR is to support the [InternLM2](https://huggingface.co/internlm/internlm2-20b). InternLM2 is a GQA-based transform architecture and offers effective support for long contexts of up to 200,000 characters. Now more and more users want to explore the capabilities of InternLM2 using their familiar llama.cpp API. We sincerely hope that our PR can be accepted. Thank you.
> 
> For information about InternLM2, please visit https://github.com/InternLM/InternLM,
> 
> related issue: #3133, #3551, #4360, #5031
> 
> cc: @yhcc, @vansinhu, @sunpengsdu

Do you intend to support Plora_A and B tensors ?

File: llama.cpp (hunk 1 above)
Comment (SolenoidWGT) [reply llama.cpp]: Thanks for your suggestion! In terms of deciding the value for `add_dummy_prefix`, I went through the SentencePiece code and found that the tokenizer's `add_dummy_prefix` value is stored [there](https://github.com/google/sentencepiece/blob/53de76561cfc149d3c01037f0595669ad32a5e7c/src/builtin_pb/sentencepiece_model.pb.h#L1315). And we can get the value from `tokenizer.model` through the python interface:
```python
from sentencepiece import sentencepiece_model_pb2 as model
m = model.ModelProto()
m.ParseFromString(open(TOKENIZER_PATH, "rb").read())
print(m.normalizer_spec.add_dummy_prefix)
```

So, maybe when we're doing the checkpoint conversion in convert-hf-to-gguf.py, we could consider adding a new key-value field using gguf_writer. For example, we might use something like 
```python
gguf_writer.add_add_dummy_prefix(add_dummy_prefix: bool).
```
If you're good with it, I'll make a change so that we're not tying the decision on `add_dummy_prefix` to any specific model type @ggerganov

**Comment by SolenoidWGT** [comment]
> > Dear llama.cpp developer @ggerganov:
> > Hello, this PR is to support the [InternLM2](https://huggingface.co/internlm/internlm2-20b). InternLM2 is a GQA-based transform architecture and offers effective support for long contexts of up to 200,000 characters. Now more and more users want to explore the capabilities of InternLM2 using their familiar llama.cpp API. We sincerely hope that our PR can be accepted. Thank you.
> > For information about InternLM2, please visit https://github.com/InternLM/InternLM,
> > related issue: #3133, #3551, #4360, #5031
> > cc: @yhcc, @vansinhu, @sunpengsdu
> 
> Do you intend to support Plora_A and B tensors ?
> 
> I've experimented with those a bit but my pytorch experience is low, so it's a pain applying it into ggml. The qkv lora is taken without reshape, I wasn't sure how to do it.
> 
> constants.py
> 
> ```
> class MODEL_TENSOR(IntEnum):
>     ATTN_QKV_LORA_A = auto()
>     ATTN_QKV_LORA_B = auto()
>     ATTN_OUT_LORA_A = auto()
>     ATTN_OUT_LORA_B = auto()
>     FFN_UP_LORA_A   = auto()
>     FFN_UP_LORA_B   = auto()
>     FFN_GATE_LORA_A = auto()
>     FFN_GATE_LORA_B = auto()
>     FFN_DOWN_LORA_A = auto()
>     FFN_DOWN_LORA_B = auto()
> ```
> 
> constants.py
> 
> ```
> TENSOR_NAMES: dict[MODEL_TENSOR, str] = {
> ..
>     MODEL_TENSOR.ATTN_QKV_LORA_A : "blk.{bid}.attn_qkv_lora_a",
>     MODEL_TENSOR.ATTN_QKV_LORA_B : "blk.{bid}.attn_qkv_lora_b",
>     MODEL_TENSOR.ATTN_OUT_LORA_A : "blk.{bid}.attn_out_lora_a",
>     MODEL_TENSOR.ATTN_OUT_LORA_B : "blk.{bid}.attn_out_lora_b",
>     MODEL_TENSOR.FFN_UP_LORA_A   : "blk.{bid}.ffn_up_lora_a",
>     MODEL_TENSOR.FFN_UP_LORA_B   : "blk.{bid}.ffn_up_lora_b",
>     MODEL_TENSOR.FFN_GATE_LORA_A : "blk.{bid}.ffn_gate_lora_a",
>     MODEL_TENSOR.FFN_GATE_LORA_B : "blk.{bid}.ffn_gate_lora_b",
>     MODEL_TENSOR.FFN_DOWN_LORA_A : "blk.{bid}.ffn_down_lora_a",
>     MODEL_TENSOR.FFN_DOWN_LORA_B : "blk.{bid}.ffn_down_lora_b",
> ```
> 
> constants.py
> 
> ```
> MODEL_ARCH.INTERNLM2: [
>         MODEL_TENSOR.TOKEN_EMBD,
>         MODEL_TENSOR.OUTPUT_NORM,
>         MODEL_TENSOR.OUTPUT,
>         MODEL_TENSOR.ATTN_NORM,
>         MODEL_TENSOR.ATTN_Q,
>         MODEL_TENSOR.ATTN_K,
>         MODEL_TENSOR.ATTN_V,
>         MODEL_TENSOR.ATTN_OUT,
>         MODEL_TENSOR.ATTN_ROT_EMBD,
>         MODEL_TENSOR.FFN_NORM,
>         MODEL_TENSOR.FFN_GATE,
>         MODEL_TENSOR.FFN_DOWN,
>         MODEL_TENSOR.FFN_UP,
> 
>         MODEL_TENSOR.ATTN_QKV_LORA_A,
>         MODEL_TENSOR.ATTN_QKV_LORA_B,
>         MODEL_TENSOR.ATTN_OUT_LORA_A,
>         MODEL_TENSOR.ATTN_OUT_LORA_B,
>         MODEL_TENSOR.FFN_UP_LORA_A,
>         MODEL_TENSOR.FFN_UP_LORA_B,
>         MODEL_TENSOR.FFN_GATE_LORA_A,
>         MODEL_TENSOR.FFN_GATE_LORA_B,
>         MODEL_TENSOR.FFN_DOWN_LORA_A,
>         MODEL_TENSOR.FFN_DOWN_LORA_B,
>     ],
> ```
> 
> tensor_mapping.py:
> 
> ```
>         MODEL_TENSOR.ATTN_QKV_LORA_A: (
>             "model.layers.{bid}.attention.wqkv.Plora_A",                 # internlm2
>         ),
>         MODEL_TENSOR.ATTN_QKV_LORA_B: (
>             "model.layers.{bid}.attention.wqkv.Plora_B",                 # internlm2
>         ),
>         MODEL_TENSOR.ATTN_OUT_LORA_A: (
>             "model.layers.{bid}.attention.wo.Plora_A",                 # internlm2
>         ),
>         MODEL_TENSOR.ATTN_OUT_LORA_B: (
>             "model.layers.{bid}.attention.wo.Plora_B",                 # internlm2
>         ),
>         MODEL_TENSOR.FFN_UP_LORA_A: (
>             "model.layers.{bid}.feed_forward.w3.Plora_A",                 # internlm2
>         ),
>         MODEL_TENSOR.FFN_UP_LORA_B: (
>             "model.layers.{bid}.feed_forward.w3.Plora_B",                 # internlm2
>         ),
>        MODEL_TENSOR.FFN_GATE_LORA_A: (
>             "model.layers.{bid}.feed_forward.w1.Plora_A",                 # internlm2
>         ),
>         MODEL_TENSOR.FFN_GATE_LORA_B: (
>             "model.layers.{bid}.feed_forward.w1.Plora_B",                 # internlm2
>         ),
>         MODEL_TENSOR.FFN_DOWN_LORA_A: (
>             "model.layers.{bid}.feed_forward.w2.Plora_A",                 # internlm2
>         ),
>         MODEL_TENSOR.FFN_DOWN_LORA_B: (
>             "model.layers.{bid}.feed_forward.w2.Plora_B",                 # internlm2
>         ),
> ```
> 
> in convert_hf_to_ggml.py: (skipping qkv reshape)
> 
> ```
>     qkv_pattern = r"model\.layers\.(\d+)\.attention\.wqkv"
>         for name, data_torch in model_kv.items():
>             # we don't need these
>             if name.endswith(".rotary_emb.inv_freq"):
>                 continue
>             plora_tensor = True if "Plora" in name else False
>             if re.match(qkv_pattern, name) and not plora_tensor:
> ```
> 
> That makes your lora models convertable, quantization works (though I guess that the LORA should get a higher quantization quality)
> 
> Now what would be missing is the conditional support to apply it ?

I will try to support lora for internLM2, but I am not very familiar with the lora implementation of llama.cpp, and I need some time to read the code.

**Comment by cmp-nct** [comment]
> 
> I will try to support lora for internLM2, but I am not very familiar with the lora implementation of llama.cpp, and I need some time to read the code.

Hi, I'm happy to hear that.
For context:
llama.cpp currently has lora support but it's nothing like your implementation.
llama.cpp precompiles a lora-delta into weights and creates a combined weight for inference, it doesn't do that during inference.

My pytorch is bad (I tried to avoid it and python, paying now), from a glance it looks like your xcomposer plora approach is not being applied on weights directly, also it's conditionally applied by a mask.
If I did not miss something crucial, then the additional tensors (as I experimented) are necessary and the architecture needs to have an additional pre-block in attention and feed forward.

File: llama.cpp (hunk 1 above)
Comment (ggerganov) [reply llama.cpp]: @SolenoidWGT Yes, I think this is a good solution. Please make the change

**Comment by cmp-nct** [comment]
I've put this PR together with xcomposer2 support here: https://github.com/ggerganov/llama.cpp/pull/5232
It does not handle the partial LoRa yet, my pytorch knoweldge is minimal and I struggle to follow the flow and shapes.

File: llama.cpp (hunk 1 above)
Comment (SolenoidWGT) [reply llama.cpp]: > @SolenoidWGT Yes, I think this is a good solution. Please make the change
Done.

**Review by ggerganov — APPROVED** [review]

File: llama.cpp [hunk 2]
```diff
@@ -664,12 +668,26 @@ static std::map<llm_arch, std::map<llm_tensor, std::string>> LLM_TENSOR_NAMES =
             { LLM_TENSOR_ATTN_OUT,        "blk.%d.attn_output" },
             { LLM_TENSOR_ATTN_ROT_EMBD,   "blk.%d.attn_rot_embd" },
             { LLM_TENSOR_FFN_NORM,        "blk.%d.ffn_norm" },
+            { LLM_TENSOR_FFN_DOWN,        "blk.%d.ffn_down" },
+        },
```
Comment (ggerganov) [inline llama.cpp:672]: Some of Orion's tensors have been deleted - `LLM_TENSOR_FFN_GATE` and `LLM_TENSOR_FFN_UP`. Please restore:

https://github.com/ggerganov/llama.cpp/blob/1cfb5372cf5707c8ec6dde7c874f4a44a6c4c915/llama.cpp#L654-L670

Comment (SolenoidWGT) [reply llama.cpp]: Sorry, I was careless when rabase, it has been fixed.

**Comment by sweetcard** [comment]
python convert-hf-to-gguf.py  ./internlm2-chat-20b
It doesn't work. Please check the following image:


image

**Comment by arch-btw** [comment]
I'm having the same issue as @sweetcard but on `internlm2-chat-7b`

**Comment by SolenoidWGT** [comment]
@arch-btw @sweetcard , I spent some time debugging to find the problem, and fixed it in this PR https://github.com/ggerganov/llama.cpp/pull/5305.

**Comment by sweetcard** [comment]
> @arch-btw @sweetcard , I spent some time debugging to find the problem, and fixed it in this PR #5305.

thank you for your excellent wok👍

**Comment by arch-btw** [comment]
Thank you @SolenoidWGT !
